### PR TITLE
ECIP-1077: Meta Specification Establishing Fork ECIPs (Using Pull Request development pattern)

### PR DESCRIPTION
**Pull request by meowsbits**
### TL;DR

Make ECIPs with `Type` _Fork_ a thing, where _Fork_ ECIPs should follow this process:
- A "shell" Fork ECIP is opened, fitting the [template provided](TODO). This ECIP __does not initially include `Activation Block Number` or `Features` specifications__.
- While in `WIP` or `Draft` status, one or more change sets are proposed against this document (eg. via Github Pull Requests) __modifying BOTH `Activation Block Number` and `Features` specifications__.
- ... Stuff happens; there are comments, emails, meetings, bribes, blogs, bragging, trolling, haranguing, arguing, debating, pondering, editing, compromising, constructive critique, etc...
- One change set is merged to modify the Fork ECIP, yielding a COMPLETE and UNIQUE hard fork specification document, and status is moved to `Last Call`. By virtue of the process specification provided, a Fork ECIP achieving `Last Call` may not see `Draft` nor `WIP` status again (because doing so would require a merging a change set holding INCOMPLETE specification variables). 

---

Related to and building from #223 , #222 , #221, #219 .

---

#### Discussion preamble from the author

This specification requires that Fork _Features_ and _Activation_ numbers are handled inextricably, and that once merged a Fork specification ECIP becomes finalized.

There are some important cases and implications to consider for this:
- If, say, a Fork ECIP specification is finalized (having been fully specified with Activation Block Number and Feature Set via a merged change set)... BUT subsequently a bug is found during testing causing a loss of confidence in sucessfully achieving the fork at the specified block number, then the Fork ECIP would want to be `Deferred` or `Withdrawn` or `Rejected`, and then eventually replaced with a new UNIQUE ECIP containing a revised Activation Block Number and possibly also Feature Set parameters. 

- Change sets (here, Pull Requests) are expected to take on a more considerable role for Fork specification development; proposed Feature Set + Activation schedule should be documented explicitly and then discussed relatively. Sensible use of Github's linking tools (Issue/PR references) should be used liberally, and it would probably be good establish a practical naming scheme for proposed Fork ECIP modifications (eg. `FORK-ECIP-9999 Spec: <My features> @ <My block>`)

### CODE REVIEW

File: _specs/ecip-_.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+ecip: TBD
```
Comment (sorpaas) [inline _specs/ecip-_.md]: You can use number 1077.

File: _specs/ecip-1000.md [hunk 2]
```diff
@@ -172,6 +172,7 @@ There are three types of ECIP:
...
   - **ECBP (Ethereum Classic Best Practice)** - application-level standards and conventions, including contract standards such as token standards, name registries, URI schemes, library/package formats, and wallet formats.
 - A **Meta ECIP** describes a **process** surrounding Ethereum Classic or proposes a change to (or an event in) a process. Process ECIPs are like Standard Track ECIPs, but apply to areas other than the Ethereum Classic protocol itself. They may propose an implementation, but not to Ethereum Classic's codebase; they often require community consensus; unlike Informational ECIPs, they are more than recommendations, and users are typically not free to ignore them. Examples include procedures, guidelines, changes to the decision-making process (e.g. this ECIP-1000 process document), and changes to the tools or environment used in Ethereum Classic development. *Any meta-ECIP is also considered a Process ECIP.*
+- A **Fork ECIP** describes a feature set and a block activation number for those features to become active on one or more blockchains.
```
Comment (sorpaas) [inline _specs/ecip-1000.md]: This should reference a link to ECIP-1077.

File: _specs/ecip-_.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+title: The Meta Fork ECIP: Establishing _Fork_ type ECIPs and their specifications and processes
+author: Mr. Meows D. Bits
+discussions-to: TBD
+status: WIP
```
Comment (sorpaas) [inline _specs/ecip-_.md]: When this ECIP is accepted, we need to decide whether the status should be "Active" or "Final". I believe "Active" will be an appropriate status here as this is an ECIP about the ECIP process that may still be modified later.

File: _specs/ecip-_.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+type: Meta
+created: 2019-11-29T14:17:42-05:00
```
Comment (sorpaas) [inline _specs/ecip-_.md]:
```suggestion
created: 2019-11-29
```

File: _specs/ecip-_.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+resolution: TBD
```
Comment (sorpaas) [inline _specs/ecip-_.md]: What does this field mean?

File: _specs/ecip-_.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+### Definition
+
+Fork ECIPs are defined as ECIPs specifying PROTOCOL ACTIVATION of any one or more Standards-Core track ECIP(s) at a specified activation block number.
+
+#### Standards
+
+Fork ECIPs should contain only placeholder information for ECIP-sets and block activation numbers while in `WIP` or `Draft` status. 
+
+Modification to any of `Last Call`, `Accepted`, `Final`, or `Active` statuses should be accompanied by the introduction (merge) of a single change set containing COMPLETE and UNIQUE definitions of the placeholder values.
```
Comment (sorpaas) [inline _specs/ecip-_.md]: I think Fork ECIPs should not have "Active" status. They should be "Final", as once it's accepted, no substance change should ever happen to the specification.

File: _specs/ecip-_.md (hunk 1 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: Done via https://github.com/ethereumclassic/ECIPs/pull/224/commits/6e9b361c2af084af9fcee9089cbad538b21e41d8

File: _specs/ecip-_.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+### Abstract
+
+Establishing _Fork_ type ECIPs, and their standards and processes.
+
+#### TL;DR
+
+Make ECIPs with `Type` _Fork_ a thing. _Fork_ ECIPs should follow this process:
+- A "shell" Fork ECIP is opened, fitting the [template provided](TODO). This ECIP __does not initially include `Activation Block Number` or `Features` specifications__.
+- While in `WIP` or `Draft` status, one or more change sets are proposed against this document (eg. via Github Pull Requests) __modifying BOTH `Activation Block Number` and `Features` specifications__.
```
Comment (sorpaas) [inline _specs/ecip-_.md]: Probably also states that those PRs should not be merged until the community consensus is reached?

File: _specs/ecip-_.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+#### Alternatives considered
+
+Use distinct UNIQUE and COMPLETE ECIPs to describe fork specification options.
+
+However, allowing fully-formed alternative and "competing" ECIPs is logistically and practically difficult to use. 
+Currently, our most prominently used collaboration tool, Github, does not provide an accessible UI for comparing arbitrary documents or review separate arbitrary documents as a conceptual set.
+
+### Implementation
+
+A Fork ECIP should achieve `Last Call` as a change set has been successfully merged. Conversely, a Fork ECIP with non-empty specifications may not see `Draft` or `WIP` status; it may only otherwise `Active`, `Withdrawn`, or `Rejected`.
```
Comment (sorpaas) [inline _specs/ecip-_.md]: This should be "Final", not "Active". Fork ECIPs are not expected to have substance change once it's accepted.

File: _specs/ecip-_.md (hunk 3 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: Agreed; `Active` seems sensible to me too.

File: _specs/ecip-_.md (hunk 8 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: Resolved with https://github.com/ethereumclassic/ECIPs/pull/224/commits/b00bdcacc07bae5436a1444f2c781c8f49508fce

File: _specs/ecip-_.md (hunk 6 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: Resolved with https://github.com/ethereumclassic/ECIPs/pull/224/commits/b00bdcacc07bae5436a1444f2c781c8f49508fce

File: _specs/ecip-_.md (hunk 5 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: No idea. Resolved with https://github.com/ethereumclassic/ECIPs/pull/224/commits/aaf75959895341c82e900081f35611886d055fe9

File: _specs/ecip-_.md (hunk 4 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: Resolved with https://github.com/ethereumclassic/ECIPs/pull/224/commits/2c40819dc9f480632a74f267e9bb828dd47657aa

File: _specs/ecip-1000.md (hunk 2 above)
Comment (meowsbits) [reply _specs/ecip-1000.md]: Resolved with https://github.com/ethereumclassic/ECIPs/pull/224/commits/b9478e610fdf1ae082ec32c0d9d4408ff5ced481

File: _specs/ecip-_.md (hunk 3 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: Resolved with https://github.com/ethereumclassic/ECIPs/pull/224/commits/b00bdcacc07bae5436a1444f2c781c8f49508fce

File: _specs/ecip-_.md (hunk 7 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: I would prefer not to mention or even begin to tangle with that idea. I'm interested here only in tangible, describable mechanics.

(Including that idea might provoke questions like "What is that?" ...Which I would prefer to leave as a dubious project for another very very brave soul to tackle in a different place at a different time.)

Comment (sorpaas) [reply _specs/ecip-_.md]: @meowsbits Yeah I understand. But what should editors do when multiple changesets are present? I just think no matter what we choose, the process specification must state this clearly.

Comment (meowsbits) [reply _specs/ecip-_.md]: Let me think about this. I will sleep on it and get back to you tomorrow.

Comment (meowsbits) [reply _specs/ecip-_.md]: I'm leaning toward keeping this document as "mechanic" as possible; to describe only "the physics" of ECIP process events. This proposal is intended to be structural, and rooted in material patterns, and to mostly just formalize an existing practice with a few minor (but important, in my opinion) tweaks. ... :smile_cat: I want to help program and code the event-based system; but not to provide the type definitions or factory for providing those events... yet )). 

Pull requests are getting merged already, somehow. ECIPs are changing status already, somehow. Human "consensus"/decision-making (either as a potential specification or as practical tradition) has important impacts for this repository's specifications including, but certainly not limited to, this one. 

If you feel strongly a need to specify and document how "human consensus" does or should happen in a given case, then I would suggest we do so in a separate concern.

I am -- as always -- open to further discussion, especially if you would (please :heart_decoration: ) propose a change set we can use a concrete reference point.

Comment (meowsbits) [reply _specs/ecip-_.md]: (And let's leave this open -- and see if anyone else has any other ideas ;)

**Comment by meowsbits** [comment]
> If, say, a Fork ECIP specification is finalized (having been fully specified with Activation Block Number and Feature Set via a merged change set)... BUT subsequently a bug is found during testing causing a loss of confidence in sucessfully achieving the fork at the specified block number, then the Fork ECIP would want to be Deferred or Withdrawn or Rejected, and then eventually replaced with a new UNIQUE ECIP containing a revised Activation Block Number and possibly also Feature Set parameters.

This implies that as discussion and review and popularization of proposed specification change sets show narrowing in apparent probable outcomes (in the minds of developers and protocol-managing project owners) that material implementations should be underway and enduring analysis and testing so that adequate and informed block numbers and any specification parameter details can be shaped and inform the developing proposals. 

In my mind this is a healthy approach: It encourages active and specification-participating research and design, prototype implementations, team coordination, and stakeholder outreach as a necessary part of the process of developing acceptable Fork ECIPs. This might work to facilitate and foster collaboration between teams and professional contexts and advocates and critics alike, while also hopefully continuing to build and prove rubrics for analysis and evaluation of proposals.

[Rough consensus (wikipedia)](https://en.wikipedia.org/wiki/Rough_consensus) is [used](https://github.com/ethereumclassic/ECIPs/search?q=rough+consensus&unscoped_q=rough+consensus) in ECIP Meta specifications. From the Wikipedia article:

> The phrase is often extended into the saying "rough consensus and running code"

In my mind, this proposal encourages a nudge in this direction, which I think has positive and constructive potential for framing our Fork ECIP processes.

File: _specs/ecip-1077.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+### Abstract
+
+Establishing _Fork_ type ECIPs, and their standards and processes.
+
+#### TL;DR
+
+Make ECIPs with `Type` _Fork_ a thing. _Fork_ ECIPs should follow this process:
+- A "shell" Fork ECIP is opened, fitting the [template provided](TODO). This ECIP __does not initially include `Activation Block Number` or `Features` specifications__.
+- While in `WIP` or `Draft` status, one or more change sets are proposed against this document (eg. via Github Pull Requests) __modifying BOTH `Activation Block Number` and `Features` specifications__.
+- ... Stuff happens; there are comments, emails, meetings, bribes, blogs, bragging, trolling, haranguing, arguing, debating, pondering, editing, compromising, constructive criticiziing, etc...
```
Comment (meowsbits) [inline _specs/ecip-1077.md]: Just posting a potential revision that I think might fit @sorpaas' intentions w/r/t to discussion at https://github.com/ethereumclassic/ECIPs/pull/224#discussion_r352308946.

```suggestion
- _Rough consensus_ (keyword: see [definition](https://github.com/ethereumclassic/ECIPs/blob/c87f1828bd3d13ae5505f229207b65079a2de915/_specs/ecip-1000.md#L205) and [uses](https://github.com/ethereumclassic/ECIPs/search?q=rough+consensus&unscoped_q=rough+consensus)) is achieved and a single change set it explicitly approved by [ALL acting ECIP Editors](https://github.com/ethereumclassic/ECIPs/blob/c87f1828bd3d13ae5505f229207b65079a2de915/_specs/ecip-1000.md#ecip-editors). 
```

File: _specs/ecip-_.md (hunk 7 above)
Comment (meowsbits) [reply _specs/ecip-_.md]: @sorpaas See https://github.com/ethereumclassic/ECIPs/pull/224#discussion_r352338257 for a draft meant to show a change that might fit your intention here.

**Comment by bobsummerwill** [comment]
I don't strongly oppose this, @meowsbits.
Wonder if we need to be this explicit?

I like @shanejonas's suggestion here too, and think it might be enough:
https://github.com/ethereumclassic/ECIPs/pull/221

I am just a little wary of OVER constraint in reaction to a series of silly events which, were they to happen again, could just be headed off socially.

**Comment by meowsbits** [comment]
@bobsummerwill I understand you concern.

Here's why I think being explicit is important:
- One of the opposing argument sets that came up in this latest flurry is "This is the way we've always done it" vs. "This is what the procedure should technically be". This attempts to _clarify_, not constrain, a hopefully healthy and accessible pattern for collaboration. 
- All creative and especially collaborative constructive projects need constraints in order to structure and direct growth. This intends to serve that purpose, offering a concrete structural framework within which any _any_ specification proposals can fit.
- I don't intend this to be only reactionary (in light of the latest flurry of debates). I intend this to actually be a constructive specification for how we can build better Forks going forward, albeit informed with some degree of hindsight wisdom/bruises.
- `could just be headed off socially`. I agree that this is an optimistic and ideal scenario. But this ideal depends on good and available leadership and mutual collaborative attitudes, which can not be taken for granted in an intentionally decentralized an decentralizing community. Relying on the availability of certain social dynamics and/or certain personalities in this space is a not a robust a system. 
- `a series of silly events `. This may be. But aside from "silly bickering" we are now looking at a maybe-seriously-silly Fork specification (either 1061 or 1072) which potentially gives rise to an important security flaw (https://github.com/ethereumclassic/ECIPs/issues/217#issuecomment-560102235), which arguably is, at least in part, caused by a failure to understand the proposed set of included ECIPs as a whole.

**Comment by bobsummerwill** [comment]
Great arguments, @meowsbits.
Makes sense to me to clarify in this way.
Let's do it.
Thanks for the extra thought process.

**Comment by meowsbits** [comment]
I'm not interested in pursuing this proposed change anymore, and have moved the `status` to `Withdrawn`. @ethereumclassic/ecip-editors please review for the merge for the sake of _documentation_.

**Comment by q9f** [comment]
Merging for the sake of documentation.
